### ISSUE TITLE
Op Support

### ISSUE THREAD

**Issue by saienduri** [nod-ai/SHARK-Turbine#215]
Tracking model burndown for Gen-AI models and variants that we seek to be serving via Turbine.

- [ ] Dynamic shaped llama2
  - [x] #153
  - [ ] Int4 Quantization with parameters from llama.cpp
  - [ ] Meeting latency requirements
- [ ] SHARK Model Porting
  - [ ] #152 (@aviator19941 )
    - [ ] Unet
    - [ ] VAE
    - [x] CLIP
    - [ ] Controlnet support (@gpetters94 )
    - [ ] LoRA support
  - [ ] https://github.com/nod-ai/SHARK/issues/1931 (@monorimet )
  - [ ] SHARK TANK -- in progress: https://github.com/nod-ai/SHARK/pull/1904 (@monorimet )

- [ ] Priority op requests
  - [ ] #210
  - [ ] #110 

- [ ] General torch-mlir op support
  - [ ] torch.aten.kthvalue
  - [ ] torch.aten._fft_c2c
  - [ ] torch.aten.fmod.Tensor
  - [ ] torch.aten.affine_grid_generator
  - [ ] torch.aten._fft_c2r
  - [ ] torch.aten.histc
  - [ ] torch.aten.poisson
  - [ ] torch.aten.logit
  - [ ] torch.aten._linalg_solve_ex
  - [ ] torch.aten._pdist_forward
  - [ ] torch.aten.linalg_cholesky_ex
  - [ ] torch.aten.triu_indices
  - [ ] torch.aten.lgamma
  - [ ] torch.aten.prod
  - [ ] torch.aten.all.dim
  - [ ] torch.aten.resize_as
  - [ ] torch.aten._cdist_forward
  - [ ] torch.aten._sample_dirichlet
  - [ ] torch.aten.remainder.Tensor
  - [ ] torch.aten._sparse_addmm
  - [ ] torch.aten.cosh
  - [ ] torch.aten.renorm
  - [ ] torch.aten.diagonal
  - [ ] torch.aten.tril_indices
  - [ ] torch.aten.upsample_trilinear3d
  - [ ] torch.aten.linalg_cross
  - [ ] torch.aten.pixel_unshuffle
  - [ ] torch.aten.rrelu_with_noise
  - [ ] #160
  - [ ] torch.aten._sparse_coo_tensor_with_dims_and_tensors
  - [ ] torch.aten.lerp.Scalar
  - [ ] torch.aten.trace
  - [ ] torch.aten.normal.float_float
  - [ ] torch.aten.hann_window.periodic
  - [ ] torch.aten._adaptive_avg_pool3d
  - [ ] torch.aten.resize
  - [ ] torch.aten.normal.Tensor_Tensor
  - [ ] torch.aten.unfold
  - [ ] torch.aten.isinf
  - [ ] torch.aten.conv_tbc
  - [ ] torch.aten.norm.Scalar
  - [ ] torch.aten._fft_r2c
  - [ ] torch.aten._linalg_slogdet
  - [ ] torch.aten.prod.dim_int
  - [ ] torch.aten._trilinear
  - [ ] torch.aten.as_strided
  - [ ] torch.aten.reflection_pad1d
  - [ ] torch.aten.exponential
  - [ ] torch.aten.randperm
  - [ ] torch.aten.min.dim
  - [ ] torch.aten.argmin
  - [ ] torch.aten.replication_pad3d
  - [ ] torch.aten.acos
  - [ ] torch.aten.diag_embed
  - [ ] torch.aten.max_unpool2d
  - [ ] #163
  - [x] #161
  - [ ] torch.aten.replication_pad2d
  - [ ] #159
  - [ ] torch.aten.normal_functional
  - [ ] torch.aten.max_pool3d_with_indices
  - [ ] #148
  - [ ] torch.aten._weight_norm_interface
  - [ ] torch.aten.reflection_pad2d

**Comment by godot73** [comment]
The list high priority ops:

Add
AveragePool
BatchNormalization
Cast
Clip
Concat
Constant
ConstantOfShape
Conv
ConvTranspose
Depth2Space
Div
DQ + Q
Expand
Gather
Gemm (->Conv2d)
GlobalAveragePool
Identity
InstanceNormalization
LeakyRelu
LSTM
MatMul
Max
MaxPool
Mul
Pad
Pow
Reciprocal
ReduceMean
ReduceMin
ReduceSum
Relu
Reshape
Resize(bilinear-> pad + transposedwc)
Shape
Sigmoid
Slice
Softmax
Split
Sqrt
Squeeze
Sub
Tanh
Tile
Transpose
Unsqueeze

**Comment by wu-s-john** [comment]
Working on the MatMul changes right now.

**Comment by Shukla-Gaurav** [comment]
Working on Gather, LeakyRelu and Pad op.

**Comment by vivekkhandelwal1** [comment]
> Working on Gather, LeakyRelu and Pad op.

Can you please create issues for these ops, so that no one else takes that op?

**Comment by vivekkhandelwal1** [comment]
@godot73 @frafranz @renxida @fredrick-h Please create an issue for the op you're working on right now, and do the same if you take further any ops.

**Comment by AmosLewis** [comment]
> @godot73 @frafranz @renxida @frederik-h @rsuderman Please create an issue for the op you're working on right now, and do the same if you take further any ops.

@wu-s-john

**Comment by wu-s-john** [comment]
@saienduri working on these two right now:

https://github.com/nod-ai/SHARK-Turbine/issues/248
https://github.com/nod-ai/SHARK-Turbine/issues/249

**Comment by wu-s-john** [comment]
Implemented https://github.com/nod-ai/SHARK-Turbine/issues/249

Need a review for my implementation of that issue.
https://github.com/llvm/torch-mlir/pull/2647

**Comment by wu-s-john** [comment]
Working on `Slice` mapping. 

Slice is much more nuanced than the operators as there is no one-to-one mapping between. I wrote a plan on how to address. Would love to get some feedback:

https://github.com/nod-ai/SHARK-Turbine/issues/259

**Comment by rsuderman** [comment]
For a list of prioritized list of issues see the following issue:
https://github.com/llvm/torch-mlir/issues/2687

These missing operations would enable lowering `resnet50` from `onnx` to `torch`

**Comment by schnkmwt** [comment]
Started working on diag_embed op
https://github.com/nod-ai/SHARK-Turbine/issues/288

**Comment by kumardeepakamd** [comment]
Started working on torch.aten.reflection_pad1d #293

**Comment by wu-s-john** [comment]
Interested in implement avg_pool3d from torch to linalg if it's not taken.

https://github.com/nod-ai/SHARK-Turbine/issues/300

**Comment by PhaneeshB** [comment]
Updated priority list with ops required for CNN models @kumardeepakamd

**Comment by daveliddell** [comment]
I can take onnx flatten
[328](https://github.com/nod-ai/SHARK-Turbine/issues/328)

**Comment by AmosLewis** [comment]
@powderluv  please add [afalkenberg1](https://github.com/afalkenberg1)

**Comment by frederik-h** [comment]
I think we should make sure to use the "Convert to issue" button on the right of each task in the list for creating a tracking issue. Unfortunately, this only becomes visible once you move the mouse over the task. If we only add a link to the task after its description, the status won't be reflected in the task list. I have now done this for my tasks and a few others, but not all. It is also possible to add the url of an existing issue as the description of the task, but then the task list will contain the name of that issue and not the name of the operator which is less readable.

**Comment by vivekkhandelwal1** [comment]
I have created a separate issue here: https://github.com/nod-ai/SHARK-Turbine/issues/347 for tracking TorchToLinalg op support. This issue is only to track OnnxToTorch op support.

**Comment by jinchen62** [comment]
@vivekkhandelwal1 According to the triage of iree onnx tests failures, these ops are missing onnx to torch support. 

- Adagrad
- Adam
- ArrayFeatureExtractor
- Binarizer
- LabelEncoder
- LpPool
- Momentum

Seems the listed ops are ordered by priority, I'm not sure how we would prioritize these missing ops. Could you add them to the list?

**Comment by Peefy** [comment]
I have an ONNX model that uses the Unique operator that is not yet supported, can I take the issue and implement it?

**Comment by renxida** [comment]
@Peefy yes please!

**Comment by vivekkhandelwal1** [comment]
Classical ML and Training ops (not planned to be supported):

- Adagrad
- Adam
- ArrayFeatureExtractor
- Binarizer
- LabelEncoder
- Momentum

**Comment by aldesilv** [comment]
I'll take group normalization.

**Comment by archana-ramalingam** [comment]
I am working on ReduceSumSquare op

**Comment by archana-ramalingam** [comment]
I am working on  ReduceLogSum op

**Comment by NeverRaR** [comment]
Working on  GlobalMaxPool

**Comment by 123epsilon** [comment]
Working on Onnx.Multinomial

**Comment by andfau-amd** [comment]
I noticed that some ONNX operators are functions, which means that we can probably systematically expand them before conversion, instead of having to write bespoke conversions for all of them. I made an issue about this: https://github.com/llvm/torch-mlir/issues/3384. Assuming this does actually get implemented, it might be wise for people considering implementing new conversions to avoid operators that are functions, if it is desirable to avoid redundant effort.

**Comment by suryajasper** [comment]
Working on Onnx.Scatter
https://github.com/nod-ai/SHARK-Turbine/issues/708

**Comment by manupak** [comment]
Working on https://github.com/nod-ai/SHARK-Turbine/issues/717

**Comment by umangyadav** [comment]
I'll take `GlobalLpPool`

**Comment by andfau-amd** [comment]
Update on the operators that are functions thing: as of https://github.com/llvm/torch-mlir/commit/51902ec2dc6df99a87e0fee092e59b492ab04837, support for this is in the importer. This means that if there's a new op to support and the ONNX documentation says "**Function**: `true`", it _may_ be easy to support by just adding a line to the allowlist in the importer. But this isn't guaranteed to work, and it requires judgment on a case-by-case basis as to whether this approach should be used instead of modifying TorchOnnxToTorch. Also, the imported function might, for example, depend on other operators that TorchOnnxToTorch doesn't support.

See also: https://github.com/llvm/torch-mlir/issues/3464.

**Comment by jinchen62** [comment]
I'm taking CenterCropPad #741 and ReverseSequence #742

**Comment by andfau-amd** [comment]
Bernoulli might be implemented wrong: https://github.com/llvm/torch-mlir/issues/3527

**Comment by mdazz** [comment]
Hello, could someone clarify what is the rationale for picking a certain opset for a given op? For example, I see Softmax is ticked off as done here, but actually [only opset version 13 is supported](https://github.com/llvm/torch-mlir/blob/f59c01fd2fbcb5162b8147f34c68cd970f2c26a9/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp#L410). Are the supported opset versions documented somewhere I might've missed?

**Comment by AmosLewis** [comment]
> Hello, could someone clarify what is the rationale for picking a certain opset for a given op? For example, I see Softmax is ticked off as done here, but actually [only opset version 13 is supported](https://github.com/llvm/torch-mlir/blob/f59c01fd2fbcb5162b8147f34c68cd970f2c26a9/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp#L410). Are the supported opset versions documented somewhere I might've missed?

Depends on the model requirement. It means when it fixed, probably only 13 needed, then when the next models need 19, we prioritize to support 19, if not, we can just leave it there for now, and prioritize other ops not implemented yet. If no model driven, when a new op implemented, we try to support the state of art ONNX opset version. But there is a tradeoff, if new version support takes too much time, we always pick up the low handing fruit first.

**Comment by vivekkhandelwal1** [comment]
Tracker https://github.com/nod-ai/SHARK-Turbine/issues/797 for the Onnx ops failing during Torch->Linalg lowering.
